projects/plugins/jetpack/modules/sharedaddy/sharing-sources.php

```diff
@@ -864,128 +914,74 @@ public function get_name() {
-			$source_name = substr( $post_data['source_name'], 0, 200 );
-		} else {
-			$source_name = '';
-		}
-
-		// Test email
-		$error = 1;   // Failure in data
-		if ( empty( $post_data['source_f_name'] ) && $source_email && $target_email && $source_name ) {
-			/**
-			 * Allow plugins to stop the email sharing button from running the shared message through Akismet.
-			 *
-			 * @module sharedaddy
-			 *
-			 * @since 1.1.0
-			 *
-			 * @param bool true Should we check if the message isn't spam?
-			 * @param object $post Post information.
-			 * @param array $post_data Information about the shared message.
-			 */
-			if ( apply_filters( 'sharing_email_check', true, $post, $post_data ) ) {
-				$data = array(
-					'post'           => $post,
-					'source'         => $source_email,
-					'target'         => $target_email,
-					'name'           => $source_name,
-					'sharing_source' => $this,
-				);
-				// todo: implement an error message when email doesn't get sent.
-				/**
-				 * Filter whether an email can be sent from the Email sharing button.
-				 *
-				 * @module sharedaddy
-				 *
-				 * @since 1.1.0
-				 *
-				 * @param array $data Array of information about the shared message.
-				 */
-				$data = apply_filters( 'sharing_email_can_send', $data );
-				if ( $data !== false ) {
-					// Record stats
-					parent::process_request( $data['post'], $post_data );
-
-					/**
-					 * Fires when an email is sent via the Email sharing button.
-					 *
-					 * @module sharedaddy
-					 *
-					 * @since 1.1.0
-					 *
-					 * @param array $data Array of information about the shared message.
-					 */
-					do_action( 'sharing_email_send_post', $data );
-				}
-
-				// Return a positive regardless of whether the user is subscribed or not
-				if ( $ajax ) {
-					?>
-<div class="response">
-	<div class="response-title"><?php esc_html_e( 'This post has been shared!', 'jetpack' ); ?></div>
-	<div class="response-sub">
-						<?php
-							printf(
-								/* Translators: placeholder is an email address. */
-								esc_html__( 'You have shared this post with %s', 'jetpack' ),
-								esc_html( $target_email )
-							);
-						?>
-	</div>
-	<div class="response-close"><a href="#" class="sharing_cancel"><?php esc_html_e( 'Close', 'jetpack' ); ?></a></div>
-</div>
-					<?php
-				} else {
-					wp_safe_redirect( get_permalink( $post->ID ) . '?shared=email' );
-				}
-
-				die();
-			} else {
-				$error = 2;   // Email check failed
-			}
-		}
-
-		if ( $ajax ) {
-			echo (int) $error;
-		} else {
-			wp_safe_redirect( get_permalink( $post->ID ) . '?shared=email&msg=fail' );
+		// Require an AJAX-driven submit and a valid nonce to process the request
+		if (
+			$is_ajax
+			&& isset( $post_data['email-share-nonce'] )
+			&& wp_verify_nonce( $post_data['email-share-nonce'], $this->get_email_share_nonce_action( $post ) )
+		) {
+			// Ensure that we bump stats
+			parent::process_request( $post, $post_data );
```

Comment:
I think adding `wp_send_json_success` here, and a `wp_die()` at the end of the function would be a good idea. (And possibly a `wp_send_json_error` if any of the conditions fail?).

Right now this response returns the html of the page, which is odd.

Reply:
Thanks for catching this! I've addressed this in ce71883, though I have the `wp_send_json_success()` call triggered by a separate check for `$is_ajax`.